### ISSUE TITLE
gearman_client_set_server_option calling failed

### ISSUE THREAD

**Issue by lkebin** [wcgallego/pecl-gearman#59]
Hi @wcgallego, 

I'm using wcgallego/pecl-gearman extension for my php project.

I have two gearman job servers A and B. If one of job servers was down but the other was up, GearmanClient::doBackground will be failed.

I got exception: Failed to set exception option from wcgallego/pecl-gearman extension.

Check wcgallego/pecl-gearman source code
https://github.com/wcgallego/pecl-gearman/blob/7480e92c630db1e5f1a51633fcb80ffdda4bd3eb/php_gearman_client.c#L290-L314

The reason that `gearman_client_set_server_option` calling failed.

I found the similar issue report from launchpad:
https://bugs.launchpad.net/gearmand/+bug/1098413

I also submit a issue to gearman/gearmand#160.

Can you please confirm `gearman_client_set_server_option` is required ?

**Comment by wcgallego** [comment]
Hey @lkebin that's the intended functionality yes, or at least that's the functionality that's existed for a while. I would love to see it fail over to alternate servers, but that's not the case and is part of the libgearman class, not the pecl wrapper

**Comment by jeff-minard-ck** [comment]
So, I'd like to bring this up for re-opening because it bit me as well.

Here's what I think is going on. In older versions of the pecl extension, the `addServer` method would call `gearman_client_set_server_option`. This is still the case - this code has changed almost none.

However, the code for `gearman_client_set_server_option` from `libgearman` is *very* different in 1.1.8 vs. newer versions of the code. Specifically, it used to just set an option for later use. The newer version of that routine, however, calls something like this:

```
    if (gearman_success(gearman_server_option(client->universal, option)))
```

Which is, unless I'm very wrong, actually causing the client/worker to connect to the server to set the "exceptions" option.

This behavior can trigger the exception `Failed to set exception option` if a server is added and does not exist.

In addition, this connection during addServer() is a clear regression of the promise `Adds a job server to a list of servers that can be used to run a task. No socket I/O happens here; the server is simply added to the list.`

**Comment by wcgallego** [comment]
Hey @jeff-minard-ck, 

Reopened here. I need to re-read this code and the change in libgearman. If I'm understanding you correctly, though, it sounds like the difference lies within the core libgearman code, right?

Pulll request #60 from @lkebin removes the exception throwing, but I'd like to look at the edge cases here first. In my head, it makes more sense to keep this closer to the underlying library rather than the pecl code wrapping around it. Removing this exception breaks expected functionality that has existed for a while. I would think creating an issue for libgearman would be the way to go, as it is breaking the contract.

I'm not closing this yet because I want to play around/think through this some more, but that's my current line of thinking. Feedback welcome.

**Comment by wcgallego** [comment]
Likewise, it looks like there's been work on it in the core lib:

https://github.com/gearman/gearmand/issues/160#issuecomment-386959439

@jeff-minard-ck, give that change a look at and see if it solves your issues. Will keep this open in the mean time

**Comment by jeff-minard-ck** [comment]
Oh, I fully agree the proposed "fix" of just removing this isn't actually what you want to do. Removing that C call makes the callbacks for exception handling in php never trigger, since the option wouldn't ever get set on the server-side. *Personally*, I'm not using exception callback handling, so [removing the calls](https://github.com/jeff-minard-ck/pecl-gearman/commit/d923652a3cabd176e6b7b332372b4cd3b59f59db) works for me, but it's not the "proper" fix.

The linked fix looks like it's attempting to go in the direction -- store the _desire_ for that server-side option and push it on first connect (which, as an aside, is what 1.1.8 use to do). 

However, looking at even the most [recent master branch version](https://github.com/gearman/gearmand/blob/master/libgearman/client.cc#L1802), the `gearman_client_set_server_option ` code is still being called, which does trace through to actually connecting to servers and setting the option.

You're probably right that this is broken upstream, more so than in the PECL extension.

**Comment by bmeynell** [comment]
I've also encountered this behavior and has been on my list to investigate. Simply put, my experience has been that if a `gearmand` instance is _not_ running, but _is_ included in `addServers()`, the `Failed to set exception option` Exception is thrown in the Worker(s). I agree with @jeff-minard-ck's sentiments:

https://github.com/wcgallego/pecl-gearman/issues/59#issuecomment-388212211:
> In addition, this connection during addServer() is a clear regression of the promise Adds a job server to a list of servers that can be used to run a task. No socket I/O happens here; the server is simply added to the list ().

https://github.com/gearman/gearmand/issues/160#issuecomment-388401293:
> More problematically, since the PECL code would detect a failure in setting the server side option and throw an exception, the addServer() php call now generates exceptions when addServer()ing instances which are down. This is not the intended functionality.

I'm currently confused though: Based on https://github.com/gearman/gearmand/issues/160#issuecomment-388437137, it looks like core `gearmand` is _not_ the point of contention.

**Comment by wcgallego** [comment]
to keep the convo going here, @jeff-minard-ck had a solid investigative thread here with one of the maintainers of gearmand: https://github.com/gearman/gearmand/issues/160#issuecomment-388878374

tl;dr - the change has been there for a couple years, but with pecl-gearman lagging behind in updates until I picked it up with PHP 7, it diverged. Other languages have picked this up and it seems like the pecl lib should do the same.

I'm going to take a look elsewhere to see how this has been adapted and see if I can implement something similar here. Thanks for patience in this, everyone!

**Comment by wcgallego** [comment]
Reading/Thinking through this some more:

Quick sum up: Looks like when it was originally created, addServer/addServers didn't create a connection to the server as we've noted in the comments above. So it didn't cost much to include it there and it frees up the PHP code from having to make another reference to this upon connection. That's no longer the case with @keyurdg's fix (https://bugs.launchpad.net/gearmand/+bug/1098413) from 2013. We need this call to gearman_client_set_server_option so that if we want an exception thrown, it can be handled in whichever way we want (try/catch).

I have a few options:
1) move the gearman_client_set_server_option calls to all places we make a call to the server. This seems pretty hairy, as it's easy to either miss locations or if we add new functions it'd be easy to forget to include this. We included it in addServer(s) functions to simplify avoiding this problem

2) We create a new function for setting this up. This would break backwards compatibility at the cost of removing the connecting from the addServer(s) function. It would also be another step in PHP code to add in.

3) I add another param (something like "ignore_exceptions") to the addServer(s) signatures making this set up optional, but calling gearman_client_set_server_option by default. This would allow BC to remain without any further changes and give users the option to ignore the exceptions if they really want to while delaying the initial connection to the server until work is actually done, fulfilling the promise in the documentation of this function. The callback wouldn't work until this is turned on, but we could then explicitly set this server option if the callback is called later.

I'm inclined to go with the third option. It's the least invasive, maintains BC and while documentation needs updating (it will no matter what), it should bring it closer to expectations.

Here's a diff of what I was thinking:
https://gist.github.com/wcgallego/baa15bb77722db892f61c5e4005d62f2

Also, would this same logic be applicable to the worker side? It too looks at gearman_worker_set_server_option when adding servers:
https://github.com/wcgallego/pecl-gearman/blob/master/php_gearman_worker.c#L296

Would love any and all feedback here.

**Comment by jeff-minard-ck** [comment]
Option 3 -- but, to be clear, the name should probably be `setupExceptionHandler` making the signature something more like:

```
public bool GearmanWorker::addServer ([ string $host = 127.0.0.1 [, int $port = 4730 [, bool $setupExceptionHandler = true ]]] )
```

_(Default true maintains backwards compat, but incurs the connection.)_ Adding a client/worker option for `$gm->turnOnExceptionHandlers()` from option 2 for full coverage would probably cover all the bases. 

With that setup I can have my PHP code `addServer` all day long and, at the start of a submission/job-processing, call `turnOnExceptionHandlers()` if I need it. That might be a bit more expensive since it could be highly repetitive (unless you start tracking that this was set in the C code, making the call in php a no-op).

Here's why I'm for this option -- my work uses gearman...rather extensively. We have a cluster of well over 200 gearman daemons (for many various reason not worth getting into here, but I assure you it's with good reasons). When we submit jobs through workers, client, web code, etc, we can end up calling a lot of `addServer()`s because the cost was always zero. If we go down the path where this _does_ incur a connection, **that** behavior change would be problematic for us.

Finally, I think this is the right solution because setting this option, IMO, is the outlier. Exception callbacks are only relevant if you're submitting "foreground" jobs using a client. To that point, I'm not even sure how it's useful to set this option as a worker -- people don't start a client to "submit an exception" so a worker wouldn't really ever trigger an exception callback. 

Regardless, while I can't speak for the world, foreground jobs have been an _incredibly_ rare use case considering you have much better options out there (like, I dunno, just a regular'ol http request to a load balancer). The number of people likely even taking advantage of this, I would wager, is very small.

_(Yes, based on the code I've seen, this problem is identical to worker/client addserver/addservers calls)_

**Comment by wcgallego** [comment]
Good food for thought. Thanks for the input. Code reviews ftw.

BTW, I'm curious as to your set up, now that you mention 200 daemons in use (consider me nerd sniped). We're doing hundreds of millions of jobs at Etsy (of various complexity and networking) across a couple dozen worker nodes and 4 GearmanD instances (2 pairs doing MySQL persistence with replication). If you're interested in sharing and talking through ideas, feel free to hit me up off thread.

**Comment by jeff-minard-ck** [comment]
On the patch:

```diff
-ZEND_BEGIN_ARG_INFO_EX(arginfo_oo_gearman_client_add_server, 0, 0, 2)
+ZEND_BEGIN_ARG_INFO_EX(arginfo_oo_gearman_client_add_server, 0, 0, 3)
        ZEND_ARG_INFO(0, host)
        ZEND_ARG_INFO(0, port)
 ZEND_END_ARG_INFO()
```

I'm *very* new to C/PHPEXT, but wouldn't each of these need another `ZEND_ARG_INFO(0, ignore_exception)` like line?

**Comment by wcgallego** [comment]
Hey @jeff-minard-ck, you're totally right. That's what I get for being slapping a diff together quickly :)

Sorry for radio silence in moving on this - over 2 weeks! I had a tweak on my local VM but writing up tests to compliment it and I was scratching my head on expected functionality behind it. I'll try to put my thoughts down here in this thread for more input when I can.

**Comment by jeff-minard-ck** [comment]
No worries; I'm compiling off a fork that's simply removed the `gearman_client_set_server_option` in the `addServer` for now. Since my use cases doesn't cover foreground jobs _nor_ handling exceptions in those, I'm safe to skip the option setting.

**Comment by wcgallego** [comment]
Hey All,

Another update. Was working through tests and chewing on a problem. It looks like gearman_client_add_server(s) also connects to the remote server. This is handled as a warning as is, which can be suppressed, but still violates the rule that no conn is made in the addServer call.

This doesn't immediately exclude the functionality mentioned above, and I do have that working with tests. Wanted to make mention of it, though, all the same. This starts getting a little hairier - do we store the host/port and do the addServer at a later time? That seems less useful. At that point, addServer(s) doesn't do much at all!

I'll have a PR ready for tomorrow I'll post. Feedback welcome

**Comment by wcgallego** [comment]
Ok! I'm on a real journey with this bug :)

The previous comment was because I was using a dummy host that wouldn't resolve to anything. addServer/addServers does work to try to resolve the host but not explicitly connect to the GearmanD server. Instead, I'm using localhost (a valid hostname) and an invalid port (4731) to test. It's not perfect, but it'll do for now to get this moving!

I'll be pushing a new tagged version alongside, but will leave this comment open for a bit so any comments/suggestions can come in. To all of you, thanks for the feedback and again the patience here, I know it took a while to get this fix in!

**Comment by wcgallego** [comment]
Going to close this up. If there are still standing issues here I haven't covered, please reopen and we can look into it again! Thanks again everyone

**Comment by fishfacemcgee** [comment]
@wcgallego It doesn't seem that the new arguments are actually optional, per #61. Can you verify?

**Comment by wcgallego** [comment]
Hey @fishfacemcgee, per your question in #61, it wasn't closed to inactivity (that's way too short a turn around!).

It is indeed optional. It's behind the pipe, which is how in pecl extensions you define optional params (the "b" here: https://github.com/wcgallego/pecl-gearman/blob/master/php_gearman_client.c#L305)

I also have a test in which I only pass in one param.
https://github.com/wcgallego/pecl-gearman/blob/master/tests/gearman_client_021.phpt#L29

This bit of code is working for me:
```
<?php
$client = new GearmanClient();
$client->addServers('localhost,localhost');
echo "done!" . PHP_EOL;
```

Note: The docs aren't yet updated (http://php.net/manual/en/gearmanclient.addservers.php) because I was waiting to close out this ticket first (just a few minutes ago!) before committing to rewriting docs and potentially confusing folks. Will get to that as well soon

If you're able to replicate this, please make sure you're on the most recent version of the repo and submit an example of the failure. Thanks!

**Comment by fishfacemcgee** [comment]
@wcgallego The issue I'm seeing is with `addServer`, as opposed to `addServers`. Let me double-check the issue and I'll get back to you with a proof of concept.

**Comment by wcgallego** [comment]
@fishfacemcgee GearmanClient::addServer() is working ok for me too:

```
<?php

$client = new GearmanClient();
$client->addServer('localhost');
echo "done!" . PHP_EOL;
```

Note: GearmanClient::addServer takes 3 params, 2 of which are optional - the string for the hostname (required), a port (optional) and the flag

**Comment by fishfacemcgee** [comment]
Ok, the issue appears to be related to the Dependency Injection library, as direct calls to `GearmanClient::addServer` appear to work just fine under 2.0.4.

**Comment by wcgallego** [comment]
Great! If the up to date version is working, going to keep this closed. Thanks for confirmation

**Comment by fishfacemcgee** [comment]
@wcgallego I'm not super familiar with the internals of the DI library, but I noticed in [the changeset for the Gearman Extension](https://github.com/wcgallego/pecl-gearman/commit/d948be1a937b1ebbfd5ad5ab84214398ec4ad777#diff-34dd01eca8e698842564e8a413addf9aR258) that a comment wasn't added for the new argument. I'm also not familiar with PHP Extension internals, but is that comment necessary for reflection or other such features that may be necessary by a DI library to work?

**Comment by wcgallego** [comment]
hey @fishfacemcgee, I'll keep looking into it, but the fastest way to resolve this is if you can show me an example of code that's failing with clear indications of the problem.

**Comment by wcgallego** [comment]
OK, Got something. Required # of params was indeed wrong:

```
<?php

$method_ref = new ReflectionMethod('GearmanClient', 'addServer');

print_r($method_ref);

echo "Number of parameters: " . $method_ref->getNumberOfParameters() . PHP_EOL;
echo "Number of *required* parameters: " . $method_ref->getNumberOfRequiredParameters() . PHP_EOL;

$method_ref2 = new ReflectionFunction('gearman_client_add_server');

echo "Number of parameters: " . $method_ref2->getNumberOfParameters() . PHP_EOL;
echo "Number of *required* parameters: " . $method_ref2->getNumberOfRequiredParameters() . PHP_EOL;
```

In 2.0.4, the required # of params is wrong. It's defined in php_gearman.c here:
https://github.com/wcgallego/pecl-gearman/blob/master/php_gearman.c#L352-L374

I've got it fixed, will push out shortly

**Comment by wcgallego** [comment]
@fishfacemcgee please try either the master branch or tag gearman-2.0.5, I *think* that should solve your problem. If not, code sample demonstrating! That would be immensely helpful. Thanks

**Comment by fishfacemcgee** [comment]
@wcgallego I built the extension locally. It's still failing, but there's a new error that makes it especially clear the issue is with the Dependency Injection library I'm using. In case you're curious, here's the error:

```
Fatal error: Uncaught DI\Definition\Exception\DefinitionException: Entry "GearmanClient" cannot be resolved: The parameter "enableExceptionHandler" of addServer() has no type defined or guessable. It has a default value, but the default value can't be read through Reflection because it is a PHP internal class.
```

That error is new as of 2.0.5 as it was previously having issues fully parsing `enableExceptionHandler`, so I think 2.0.5 is generally a fix/improvement over the 2.0.4 behavior, and thanks to it I now know for certain that the workaround I used is mandatory.

Thank you so much for being so responsive on this issue and digging in when I had very limited debug/repro steps.

**Comment by wcgallego** [comment]
What Dependency Injection Lib are you using? I'm trying to see if I can replicate it using ReflectionMethod on the class and getType from the params of that reflection, but it's not returning the type for any of the params, let alone the last one. That said, I'm also not able to do that with standard lib PHP functions either.

Otherwise, not sure how else to approach this, though I'm glad we've fixed the required params for these functions. Thanks for the help there!

**Comment by fishfacemcgee** [comment]
@wcgallego I'm using [php-di](https://github.com/PHP-DI/PHP-DI). To be clear, I don't think there's anything left to fix on the Gearman Extension side of things. We were auto-wiring the GearmanClient class, and PHP-DI has [an issue logged](https://github.com/PHP-DI/PHP-DI/issues/470) where it doesn't truly support auto-wiring Internal Classes (which GearmanClient would qualify as). It just didn't manifest as an issue in my codebase until the new argument was added. We've now taken a factory-ish approach (similar to what's shown in the PHP-DI issue) and everything works fine both in PECL-Gearman < 2.0.4 and PECL-Gearman >= 2.0.4.

**Comment by microlancer** [comment]
Is there any plan to also support this for `addServer()` and `addServers()` for the `\GearmanWorker` class as well? I was able to migrate my code to use the 3rd parameter for `\GearmanClient` and avoided the `GearmanException: Failed to set exception option` for that case.

However, for `\GearmanWorker` I'm still getting an error.

```
PHP Fatal error:  Uncaught GearmanException: Failed to set exception option in Gearman.php:33                                                                  
Stack trace:                                                                              
#0 /home/dev/app/Util/Gearman.php(33): GearmanWorker->addServer('127.0.0.1', 4730)
```
And trying to add the 3rd parameter (set to false), I am getting:
```
Warning: GearmanWorker::addServer() expects at most 2 parameters, 3 given in /home/dev/app/Util/Gearman.php on line 33
```
gearman extension version => 2.1.0
libgearman version => 1.1.19.1

Thanks for any input!

**Comment by wcgallego** [comment]
Hi Tim, this thread is closed and the project has been shifted to official PHP channels. Please see the top of the README
